README.md
# README
This project was based off an software engineering interview question given to me to complete.

## Table of Contents

1. [Description of problem](#description-of-problem)
1. [Set up](#set-up)
1. [Usage](#usage)

## Description of problem:

Create a restful endpoint that intakes a base64encoded FLAC audio file and stores the transcribed text into a DB table along with the time the request was made.
Google Speech API docs: https://cloud.google.com/speech-to-text
This is to be written in a framework you are familiar with, however we require unit tests and expect to see regular commits to demonstrate your approach to the task. 
Please submit this via a github repo link, with a README file. Include project setup steps and instructions to connect to your endpoint.


After completing this task, I have then further extended the project to include the ability to fetch the data stored in the database when a get request is recieved. Aswell as creating a database, if one is not already in the same directory as Project.js. I have written this project in javascript, implementing a sqlite3 database.

## Set up:

### Install Dependinces 
First, Node JS will be required. This can be installed at https://nodejs.org/en/download. 
Once Node JS is set up, in the command line, install the Google Cloud Speech, Express and SQLite3 dependencies. This is done by running the commands, npm install express, npm install @google-cloud/speech, npm install sqlite3.

If your computer does not come with SQLite3 installed, you will need to install it.

### Authorizing Google API 
For this project, you will need to enable the Google Speech-to-Text API in the Google Cloud console. Once you have selected a project and linked it to a billing account, you can enable the Speech-to-Text API. Go to the Search products and resources bar at the top of the page and type in "speech". Select the Cloud Speech-to-Text API from the list of results. You now must link one or more service accounts to the Speech-to-Text API. Click on the Credentials menu item on the left side of the Speech-to-Text API main page. Make sure billing is enabled for the API and that the project has at least one service account. A private key for that google service account is required. This will need to be stored in a file called key.json in the same directory as Project.js. Instructions for this are listed below:

You need to obtain credentials in the form of a public/private key pair. These credentials are used the code to authorize service account actions within your app.
To obtain credentials:
In the Google Cloud console, go to Menu menu > IAM & Admin > Service Accounts.
Go to Service Accounts
Select your service account.
Click Keys > Add key > Create new key.
Select JSON, then click Create.
Your new public/private key pair is generated and downloaded to your machine as a new file. Save the downloaded JSON file as key.json in your working directory.

### Files to download
Download package-lock.json, package.json and Project.js and place them in the same folder. 

## Usage:
You can connect to the endpoint of the project using an application like postman and connecting to http://localhost:8040/.

### Push requests

To have an audio file converted into text, send a post request to http://localhost:8040/get_speech_as_text with the body of that request containing a JSON file with an Object "audio" containing the based 64 encoded version of the flac audio file in the data property of the audio object. This can be created by using the conversion tool, https://base64.guru/converter/encode/audio/flac, which will convert your flac file to a correctly formatted json file. This will connect to the Google Speech to Text API and will store the text, inputted audio and date of the post request in a database titled SpeechToText.db (If this database does not already exist, it will create it). 

### Get requests

If you wish to get the results of your post requests then you can send either one of two get requests to http://localhost:8040/. 

Sending a get request to http://localhost:8040/get_text_results will return the text in the audio and the date of the text being added in a JSON file. 

Sending a get request to http://localhost:8040/get_text_results_and_audio will return the text in the audio, the audio inputted and the date of the text being added in a JSON file.

ChangeLog.md
# Change log

All notable changes to this project are documented in this file.

## 18/07/2023, 11:40am
Added small code fixes and added PR template.

## 17/07/2023, 15:58pm
No functionality added. Cleaned up code.

## 17/07/2023, 12:43pm
Added functionality to add the correct tables to the database when it connects and fucntionailty for get requests to be recieved that will return the text, inputted audio (Depending on which get request), and the date of every record in the database. Fixed bugs of data being inserted before tables were created.

## 17/07/2023, 10:29am
Added the functionality to check if the database already exists in the working dictionary, if it doesn't it then creates a database and the appropriate tables. Fixed bugs where the promise of data being inserted wasn't being resolved before moving on.

## 14/07/2023, 4:23pm
Added validation to check that audio passed in the push request is base 64 encoded.

## 14/07/2023, 4:07pm
Added the functionality for the program to connect to an SQLite3 database and store the result from the Google Speech to Text API in the database. Added exception handling.

## 14/07/2023, 12:10pm
No functionality added. Cleaned up code.

## 14/07/2023, 12:00pm
Implemented the express dependancy. Created a RESTful endpoint. Allowed for user to send a push request to http://localhost:8040/get_text reading the base 64 encoded audio from a JSON file in the request's body. This audio passed in the push request is then passed into the Google Speech to Text request.

## 13/07/2023, 3:41pm
No functionality added. Cleaned up code, seperating it into more precise, smaller functions.

## 13/07/2023, 3:28pm
Added the functionality to accept a file path of a flac recording as a command line argument, read that file and convert it to a base 64 encoded string which is then passed into the Google Speech to Text API request. Removed the need for Google Buckets.

## 13/07/2023, 3:12pm
Added the functionality to read a user's google cloud service account key, and submit a request to the Google Speech to Text API, passing the address of a fixed flac file in a Google Bucket.

Project.js
const sqlite3 = require('sqlite3')
const express = require('express');
const app = express();
const cors = require('cors');
const PORT = 8040;
const speech = require('@google-cloud/speech');
const path = require('path');
const fs = require('fs');
var base64Regex = /^([0-9a-zA-Z+/]{4})*(([0-9a-zA-Z+/]{2}==)|([0-9a-zA-Z+/]{3}=))?$/;

app.use(express.json({limit: '100mb'}));
const keyFilename = path.join(__dirname, 'key.json');
const client = new speech.SpeechClient({
  keyFilename: keyFilename
});

app.use(cors());

async function connectToDatabase(writing) {
  try{
    let database = fs.statSync("./SpeechToText.db");
    database.isFile();
    let db = new sqlite3.Database('./SpeechToText.db', sqlite3.OPEN_READWRITE);
    await createTables(db).catch(() => {return true;});
    return db;
  } catch {
    if(writing){
      let db = await createDatabase().catch(() => {return true;});
      return db;
    } else {
      return true;
    }
  }
}

async function createDatabase(){
  return new Promise((resolve, reject) => {
    let db = new sqlite3.Database('SpeechToText.db', async (err) => {
      if (err) {
          reject(true);
      } else {
        await createTables(db).catch(() => {reject(true);});
        resolve(db);
      }
    });
  })
}

async function createTables(db){
  return new Promise((resolve, reject) => {
    db.run('CREATE TABLE IF NOT EXISTS Entries(id INTEGER PRIMARY KEY AUTOINCREMENT, text TEXT, inputtedAudio TEXT, dateAdded TEXT);', [], function(err) {
      if(err){
        reject(err);
      } else {
        resolve(true);
      }
    });
  });
}

async function insertEntry(text, audioBytes, db){
  return new Promise((resolve, reject) => {
    const date = new Date();
    let dateAdded = date.toLocaleDateString();
    db.run('INSERT INTO Entries (text, inputtedAudio, dateAdded) VALUES (?,?,?)', [text, audioBytes, dateAdded], function(err) {
      if (err){
        reject(err);
      } else {
        resolve(true);
      }
    });
  });
}

async function getEntriesAsJSON(db, includeInputtedAudio){
  return new Promise((resolve, reject) => {
    if(includeInputtedAudio){
      db.all('SELECT text, inputtedAudio, dateAdded FROM Entries', (err, rows) => {
        if(err){
          reject("Failed to access Enteries in database, please check if there is an Entries table in the database");
        } else {
          resolve(JSON.stringify(rows));
        }
      });
    } else {
      db.all('SELECT text, dateAdded FROM Entries', (err, rows) => {
        if(err){
          reject("Failed to access Enteries in database, please check if there is an Entries table in the database");
        } else {
          resolve(JSON.stringify(rows));
        }
      });
    }
  })
}

async function getEntries(includeInputtedAudio){
let db = await connectToDatabase(false);
  if(!db){
    return "Failed to connect to database";
  }
  try{
    let JSONResults = await getEntriesAsJSON(db, includeInputtedAudio);
    return JSONResults;
  } catch (err){
    return err;
  }
}

async function getText(audioBytes) {
    const config = {
      encoding: 'FLAC',
      languageCode: 'en-UK',
    };
    const request = {
      audio: {content : audioBytes},
      config: config,
    };
    
    const [response] = await client.recognize(request);
    const transcription = response.results
      .map(result => result.alternatives[0].transcript)
      .join('\n');
    return transcription;
}

app.listen(
  PORT, 
  () => console.log('Its alive on https://localhost:8040')
)

app.post('/get_speech_as_text', async (req, res) => {
  const {audio} = req.body;
  if(!audio || !audio.data) {
    return res.status(418).send({message: "A FLAC audio file is required as input"})
  }
  try {
    if(!base64Regex.test(audio.data)){
      return res.status(500).send({message:"The file pass was not base 64 encoded"});
    }
    let text = await getText(audio.data);
    let db = await connectToDatabase(true);
    if(!db){
      return res.status(500).send({message:"Failed to connect to database / create tables in database"});
    }
    let successfulInsertion = await insertEntry(text, audio.data, db).catch(() => false);
    if(successfulInsertion){
      return res.status(200).send({message: "Wayhey!"})
    } else {
      return res.status(500).send({message:"Failed to insert entry into database"});
    }
  } catch (e){
    console.log(e);
    return res.status(500).send({message:"An error occurred"});
  }
})

app.get('/get_text_results', async(req, res) => {
  returnValue = await getEntries(false);
  try{
    JSON.parse(returnValue);
    res.set({'Content-Type': 'application/json'});
    res.status(200).send(returnValue);
  } catch {
    return res.status(500).send({message:returnValue});
  }
})


app.get('/get_text_results_and_audio', async(req, res) => {
  returnValue = await getEntries(true);
  try{
    JSON.parse(returnValue);
    res.set({'Content-Type': 'application/json'});
    res.status(200).send(returnValue);
  } catch {
    return res.status(500).send({message:returnValue});
  }
})

